### PR TITLE
Add auth example

### LINKED ISSUE #153: Add login / authentication example
With:
- re-usable authentication helper across pages
- session synchronization among tabs
- simple passwordless email backend hosted on `now.sh`

I think this will be hugely helpful to a lot of newcomers.

### PR DESCRIPTION
**Pull request by timneutkens**
Fixes #153

Implements the auth parts of https://github.com/iaincollins/nextjs-starter with a few differences in the client side code. In the original nextjs-starter all pages extended from a page component. I changed this to a higher order component. Added another higher order component to restrict pages, cleaned up the docs, cleaned up the code to fit `standard` and finally removed all unnecessary code.

### CODE REVIEW

File: examples/with-auth/components/logout-button.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/* global window */
+import React from 'react'
+import Session from './session'
+
+async function logout (event) {
```
Comment (davibe) [inline examples/with-auth/components/logout-button.js]: is it `logout` or `signout`

Comment (timneutkens) [inline examples/with-auth/components/logout-button.js]: To be consistent it should be signout. Good catch.

**Comment by impronunciable** [comment]
@timneutkens @rauchg wanted auth to be provided by an external service, running on now. Is it possible with this approach?

**Comment by timneutkens** [comment]
@impronunciable It's not, but I'm already working on a new implementation similar to zeit.co's login after talking with @rauchg on slack.

**Comment by gustavomanolo** [comment]
Hi there, when is this example going to merged in master?
Thanks.

**Comment by timneutkens** [comment]
@gustavomanolo most likely it's not going to be merged. Still working on the new example that doesn't use the programmatic api. Will keep you posted 👍

**Comment by gustavomanolo** [comment]
@timneutkens Thanks a lot, so meanwhile, what do you recommend to implement authentication?

**Comment by timneutkens** [comment]
@gustavomanolo This implementation is pretty solid. It can be used in production 😄 We just want to show a different example of how to implement authentication. Maybe I'll move this one into a separate repo.

**Comment by gustavomanolo** [comment]
@timneutkens Perfect, so can i run the example on https://github.com/zeit/next.js/pull/1141/files/fc067b1696400f5633bb7248b1d6f35b58ab9f81? to take it as a base.

File: examples/with-auth/components/session.js [hunk 2]
```diff
@@ -0,0 +1,204 @@
+/**
+ * A class to handle signing in and out and caching session data in sessionStore
+ *
+ * Note: We usewindow.XMLHttpRequest() here rather than fetch because fetch() uses
+ * Service Workers and they cannot share cookies with the browser session
+ * yet (!) so if we tried to get or pass the CSRF token it would mismatch.
```
Comment (sedubois) [inline examples/with-auth/components/session.js]: Would using https://github.com/developit/unfetch allow to solve this issue?

Comment (iaincollins) [inline examples/with-auth/components/session.js]: Hey interesting, it uses XMLHttpRequest under the hood and not fetch, so that seems like it should work!

Comment (impronunciable) [inline examples/with-auth/components/session.js]: I also see a typo on `Note: We usewindow.XMLHttpRequest()` (separate `use` from `window.`

File: examples/with-auth/index.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  server.use(function (req, res, next) {
+    res.setHeader('Vary', 'Accept-Encoding')
+    next()
+  })
+
+  server.listen(process.env.PORT || 3000, err => {
+    if (err) {
+      throw err
+    }
+    console.log('> Ready on http://localhost:' + process.env.PORT)
```
Comment (impronunciable) [inline examples/with-auth/index.js]: By default this runs over port 3000 (because of the `||` in line 85). move the `process.env.PORT || 3000` to a variable and use that. You can also use a template string for constructing this

File: examples/with-auth/pages/auth/error.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import Link from 'next/prefetch'
```
Comment (impronunciable) [inline examples/with-auth/pages/auth/error.js]: Use `next/link` with the `prefetch` props

File: examples/with-auth/pages/auth/not-configured.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import Link from 'next/prefetch'
```
Comment (impronunciable) [inline examples/with-auth/pages/auth/not-configured.js]: Use `next/link` with the `prefetch` props

File: examples/with-auth/pages/auth/success.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import Link from 'next/prefetch'
```
Comment (impronunciable) [inline examples/with-auth/pages/auth/success.js]: Use `next/link` with the `prefetch` props

File: examples/with-auth/pages/index.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import Link from 'next/prefetch'
```
Comment (impronunciable) [inline examples/with-auth/pages/index.js]: Use `next/link` with the `prefetch` props

File: examples/with-auth/components/with-session.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  static async getInitialProps (ctx) {
+    const session = new Session({ req: ctx.req })
+
+    let initialProps = {}
+    if (Component.getInitialProps) {
+      initialProps = Component.getInitialProps({ ...ctx, session })
+    }
+
+    const sessionData = await session.getSession()
+    console.log(sessionData)
```
Comment (impronunciable) [inline examples/with-auth/components/with-session.js]: Remove this log

**Review by impronunciable — CHANGES_REQUESTED** [review]

File: examples/with-auth/components/session.js (hunk 2 above)
Comment (iaincollins) [inline examples/with-auth/components/session.js]: Note: I only used window.XMLHttpRequest to make the linter (xo) happy, I have since changed this in my own code to tell it XMLHttpRequest is a global so it doesn't trigger a warning.

Just to clarify, the session code still needs to use XMLHttpRequest (or something that uses it under the hood, like unfetch) rather than just fetch though, as browsers don't yet let support fetch reading existing http cookies (though not by design as part of some security design, it's just not supported yet but is slated as a future feature, at least by the Mozilla team).

**Comment by impronunciable** [comment]
@timneutkens let me know when you have time to review this small issues so we can get this merged :)

**Comment by gustavomanolo** [comment]
Hi, when is this going to be added to examples?

File: examples/with-auth/components/session.js (hunk 2 above)
Comment (koutsenko) [inline examples/with-auth/components/session.js]: are you talking about "credentials" option of fetch? it allows send cookies also. or i misunderstood smthing...
```
fetch(url, {  
  credentials: 'include'  
})
```

Comment (iaincollins) [inline examples/with-auth/components/session.js]: @koutsenko Specifically fetch doesn't yet have ability to do HTTP Only cookies (i.e. the kind that can't be accessed by JavaScript, which give protection from XSS from malicious users or adverts inject JS into your code). The workalike that @sedubois suggests above looks like a good drop in replacement though.

sedubois

File: examples/with-auth/routes/auth.js [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+            throw err
+          }
+        })
+      } else {
+        User.create({email: email, token: token}, function (err) {
+          if (err) {
+            throw err
+          }
+
+          nodemailer
```
Comment (jorgeer) [inline examples/with-auth/routes/auth.js]: Shouldn't this execute whether or not the user exists? As it is now,  a new token is generated and saved for the existing user, but he is not getting the updated verification url as a token, so the old link he got will no longer work.

File: examples/with-auth/components/with-session.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import React from 'react'
+import Session from './session'
+
+export default (Component) => class extends React.Component {
+  static async getInitialProps (ctx) {
+    const session = new Session({ req: ctx.req })
+
+    let initialProps = {}
+    if (Component.getInitialProps) {
+      initialProps = Component.getInitialProps({ ...ctx, session })
```
Comment (alex-di) [inline examples/with-auth/components/with-session.js]: I believe there should be await

File: examples/with-auth/components/restricted.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+    <p><Link href='/auth/signin'><a>Please log in</a></Link></p>
+    <p><Link href='/'><a>Back to homepage</a></Link></p>
+  </div>
+)
+
+export default (Component) => {
+  const checkAuth = (props) => {
+    return props.isLoggedIn ? <Component {...props} /> : <Unauthorized />
+  }
+
```
Comment (alex-di) [inline examples/with-auth/components/restricted.js]: checkAuth fucked up props of the original component, so I added 
`if (Component.getInitialProps)
    checkAuth.getInitialProps = Component.getInitialProps
`
There must be more elegant way to do it, but it works for me

**Comment by bookercodes** [comment]
Can someone please help me understand how `localStorage` is utilised here? Having played with the [original demo](https://nextjs-starter.now.sh/auth/signin) with Chrome Developer Tools open, I have never seen `localStorage` set (it's always empty), and yet `localStorage` is referenced a lot in the code (particularly in the `session` class)

I will just add that from a Next newbie's perspective, I feel like this example, even with comments (which are a great start!), is quite hard to grok. There's a lot going on in one example, which makes it hard to follow. I think separate, stripped-down examples would work best or an accompanying step-by-step tutorial where the example evolves more gradually

**Comment by radum** [comment]
@bookercodes It is used to store the session once logged in. And in that session there info about name / email. And it can be used in the app to show it on the page. On that example it shows your name or email in the menu.

**Comment by bookercodes** [comment]
@radum Makes a lot of sense! Howcome I can't see it in local storage/session storage, though?

https://www.dropbox.com/s/rw5arp83536nqms/Screenshot%202017-06-01%2015.47.08.png?dl=0

**Comment by iaincollins** [comment]
@bookercodes That's a good question and I don't know. o_O It may be that the example needs updating to include a fix for bugs that have been addressed or the browser is being weird.

Might worth trying to sign in with oAuth on https://nextjs-starter.now.sh and seeing if it works for you there (it's a had a few fixes and improvements for robustness since then and I think works pretty well now - although emails from the live server are being spam blocked now I think so probably you can only sign in with oAuth; I need to configure the demo site to use an external SMTP server).

RE: Complexity, I agree. I've been leaving how to simplify it as something in the back of my mind for a while now. It's slightly icky as this attempt at a 'best practice' implementation requires having people include multiple components in the front end of their app - both client and server side - and wrap their head around CSRF and server side sessions support, and allowing for considerations like able to customise user models to store in whatever database they want to use.

One way to make it easier might be to create an implementation that sacrifices some of the features or security options and/or dropping server side rendering creating a simpler model that is easier to implement.

**Comment by radum** [comment]
@bookercodes I think you need to open it to see it 😁 

image

See what I mean, you need to open the LocalStorage folder and click on the link. Then you will be able to see the keys for that website.

**Comment by bookercodes** [comment]
@radum Honestly, I thought I expanded it! Just double-checked and now I see it. What a goof!

**Comment by bookercodes** [comment]
Just jotting this down for my future reference (hopefully I can expand on it), but I feel like most apps want to restrict access to multiple routes, not just one (i.e. `pages/restricted.js`). Could be worth showing how to tersely authenticate multiple routes.

**Comment by juhaelee** [comment]
It looks like this example is exposing csrf through AJAX [#L94](https://github.com/timneutkens/next.js/blob/with-auth/examples/with-auth/routes/auth.js#L94) which is unsecure no?: https://github.com/pillarjs/understanding-csrf#csrf-tokens

*I'm no expert in this field

**Comment by iaincollins** [comment]
@juhaelee

That comment is misguided, I see someone has called it out on Stack Overflow as well.

Additionally, it's not possible to use CSRF with a JavaScript SPA without providing a way to return the a CSRF to the app (whether that's via a REST endpoint or a web socket). Bear in mind requests need an HTTP Only cookie *and* the matching CSRF token for each request to perform an action.

If someone performs a request from another domain they won't be able to read your CSRF token because they will have their own session (requests from other domains wont be able to access the HTTP only cookie that contains your session token). 

You are still projected by CSRF - i.e. they can't do a remote POST request via something like JS or an form that submits an your site, triggering an action as if it were being done by a user. That's the purpose of a CSRF token.

**Comment by juhaelee** [comment]
@iaincollins i see. thanks for the detailed information! but for next.js since the first render is always on the server, couldn't we just store the CSRF token on that render and for every sequential call, use that csrf token for POST?

**Comment by iaincollins** [comment]
@juhaelee Technically you could that if you rolled your own CSRF token system, but normally CSRF tokens rotate to provide additional security - which is exactly how the CSRF module for express works - so you'd have no way making subsequent requests if you only had the first token.

File: examples/with-auth/components/session.js [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+ */
+export default class Session {
+  constructor ({req} = {}) {
+    this._session = {}
+    this._user = {}
```
Comment (juhaelee) [inline examples/with-auth/components/session.js]: this doesn't appear to be used anywhere

File: examples/with-auth/components/session.js [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+    try {
+      if (req) {
+        // If running on server we can access the server side environment
+        this._session = {
+          csrfToken: req.connection._httpMessage.locals._csrf
```
Comment (juhaelee) [inline examples/with-auth/components/session.js]: it seems that `getCsrfToken()` is always called before a POST request is made, so is this line necessary? Why store the csrfToken if every time we just update the csrfToken regardless?

it might be worth using something like axios where we can use an interceptor:
```
axios.interceptors.request.use(function(config) {
  return new Promise(async function(resolve) {
    if (config.method === 'post') {
      const csrfToken = await getCsrfToken();
      config.headers = {
        ...config.headers,
        'csrf-token': csrfToken,
      };
    }
    resolve(config);
  });
});
```

File: examples/with-auth/pages/auth/signin.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    })
+  }
+
+  render () {
+    let signinForm = <div />
+    if (!this.props.session.user) {
+      signinForm = (
+        <div>
+          <form id='signin' method='post' action='/auth/email/signin' onSubmit={this.handleSubmit}>
+            <input name='_csrf' type='hidden' value={this.props.session.csrfToken} />
```
Comment (juhaelee) [inline examples/with-auth/pages/auth/signin.js]: is this hidden input necessary with this already in place? https://github.com/zeit/next.js/pull/1141/files#diff-cba823cf5d072f066058359e8242073aR145

**Comment by eashish93** [comment]
According to this link [understanding csrf](https://github.com/pillarjs/understanding-csrf). Don't expose csrf token via ajax route. But in your code, you expose it like below:
```js
// Add route to get CSRF token via AJAX
  server.get(path + '/csrf', (req, res) => {
    return res.json({csrfToken: res.locals._csrf})
  })
```
Is it correct or that doc missing something ?

**Comment by iaincollins** [comment]
@eashish93 That post is bogus on that point, I've already been over it a above. [I've also written a long post about it on an answer in Stack Overflow](https://stackoverflow.com/questions/33574343/getting-csrf-token-via-ajax/44345692#44345692) if anyone wants an an expanded version.

If you don't want to do client side rendering then you don't need it, but if you don't want to do client side rendering ¯\_(ツ)_/¯

**Comment by timneutkens** [comment]
Closing this in favor of the new example we'll introduce in v5

**Comment by Jero786** [comment]
> Closing this in favor of the new example we'll introduce in v5

@timneutkens Hi! just a question, after I read a ton of post like [this](https://github.com/zeit/next.js/issues/153)), I conclude that it prefer to store the JWT token in cookies (Front-end) and also it's best practice to use `HTTPOnly` from the server (to not have access to get the token from JS) and `Secure` (to be sure that only will be sent by HTTPS). 

So my Backend it provide me a JWT token, I just only need to store it in a cookie. 
Also the way that check if the user is logged or not is, by a HOC, which wrap all my private pages (authenticated pages).

Questions:

* I'm going in the right direction? 
* it's really needed to use `HTTPOnly` and `Secure` a cross the pages? do you have some example?
btw, I don't need Auth0, (not needed facebook, gmail, etc user authentication)  


Thanks in advance for any help that you could bring to me! :)
